[clang][PAC] Fix builtins that claim address discriminated types are bitwise compatible

ojhunt opened the pull request:
A number of builtins report some variation of "this type is compatibile with some bitwise equivalent operation", but this is not true for address discriminated values. We had address a number of cases, but not all of them. This PR corrects the remaining builtins.

Fixes #154394

ojhunt committed:
[clang][PAC] Fix builtins that claim Address discriminated types are copyable

A number of builtins report some variation of "this type is compatibile with
some bitwise equivalent operation", but this is not true for address discriminated
values. We had address a number of cases, but not all of them. This PR corrects
the remaining builtins.

Fixes #154394

cor3ntin commented on clang/lib/AST/DeclCXX.cpp:
```diff
@@ -1438,6 +1438,10 @@ void CXXRecordDecl::addedMember(Decl *D) {
         data().StructuralIfLiteral = false;
     }
 
+    if (!data().HasTrivialSpecialMembers &&
+        T.hasAddressDiscriminatedPointerAuth())
+      data().HasTrivialSpecialMembers = true;
```
isn't that doing the opposite of what we want?
ie a discriminated field should not be trivial

ojhunt replied:
That is actually just complete nonsense, and it seems really weird that nothing broke at all.

ojhunt replied:
I think I put this in while fighting battery life last night, but I'm super annoyed that nothing got triggered warning wise (maybe need to add yet more warnings for preferred type?).

But more annoyingly I can't find a way to make it do the wrong thing - basically what this code ends up doing is saying none of the special member functions are defaulted, which is already going to be true due to the address discriminated value - *except* the addition of the default constructor which is set arbitrarily. This should make it possible to have a class that has a non-defaulted default constructor be treated as if it were trivially defaulted. But to hit the code paths that actually change behavior due to that you need to get past things like pod type checks (which always fail with address discrimination).

Annoying that something this dumb is wrong, and not caught by the compiler, and cannot be tested because the same state that leads to the error means the error cannot impact anything.

ojhunt committed:
Remove dumb code

rjmccall commented on clang/test/SemaCXX/ptrauth-triviality.cpp:
```diff
@@ -74,7 +74,7 @@ static_assert(__is_trivially_destructible(S3));
 static_assert(!__is_trivially_relocatable(S3)); // expected-warning{{deprecated}}
 //FIXME
-static_assert(__builtin_is_cpp_trivially_relocatable(S3));
+static_assert(!__builtin_is_cpp_trivially_relocatable(S3));
```
Hmm, is this right? My understanding is that, while the pre-standardization concept of trivial relocatability implies bitwise relocation, the standardized concept does not and allows for arbitrary compiler-synthesized logic as long as that logic can't cause arbitrary side-effects. I would expect address-discriminated ptrauth to not disqualify a type from that.

rjmccall replied:
I see. The idea is that it should work, and while we don't support it today, we do intend to in the future? That seems reasonable.

Wait, trivial relocation of a polymorphic object just copies the v-table pointers? That seems obviously incorrect unless there's some semantic reason relocation can only happen on complete objects.

ojhunt replied:
> Wait, trivial relocation of a polymorphic object just copies the v-table pointers? That seems obviously incorrect unless there's some semantic reason relocation can only happen on complete objects.

I also dislike it - but a bunch of people believe that it absolutely must be a memcpy. Allowing a fixup for ptrauth was a struggle as it was.

A non-zero part of me wants it to be an error if there's a constant copy size of one element - on the basis that a variable count implies that you have an array so they must all be the same type.

But yeah, blind copy of the vtable is questionable, but once again just slap a "it's UB to truncate the object" is the solution :-/

ojhunt committed:
Explicitly acknowled the impact of pointer auth on non-trivial
members

Any address discriminated values mean that the type is never trivially
copyable or movable.

ojhunt committed:
Add a .

rjmccall reviewed (approved):
This is fine with me, but it'd be nice if @cor3ntin signed off, too.

ojhunt committed:
Add comment correction from Joggn

cor3ntin reviewed (approved):
Thanks!
